### PR TITLE
Refactor Toolbar component to TypeScript

### PR DESCRIPTION
**Pull request by mike-day**
Part of #35744

## What?

Convert the `Toolbar` component to TypeScript.

## Why?

As a contribution to the ongoing effort to convert `@WordPress/components` to TypeScript.

## How?

By adding types to `Toolbar`, `@ts-nocheck` where needed for related `Toolbar` components, and refactoring of stories.

## Testing Instructions

1. Create or otherwise open a post
2. Add a block that would have some toolbar controls (e.g. `Image`)
3. Expected: The toolbar should fundamentally work the same as before:
Screen Shot 2023-01-24 at 6 38 05 PM

### CODE REVIEW

**Comment by ciampo** [comment]
Hey @mike-day , thank you so much for starting the work on this!

I will try to come back with a proper review later, but for now I just wanted to say that there will be some conflicts with PR https://github.com/WordPress/gutenberg/pull/47117 — I would recommend you wait until that's merged (hopefully within 1-2 days) and then rebase your PR on top of latest `trunk`. There will be some conflicts on the Storybook files but they should be minor :)

**Comment by mike-day** [comment]
> Hey @mike-day , thank you so much for starting the work on this!
> 
> I will try to come back with a proper review later, but for now I just wanted to say that there will be some conflicts with PR #47117 — I would recommend you wait until that's merged (hopefully within 1-2 days) and then rebase your PR on top of latest `trunk`. There will be some conflicts on the Storybook files but they should be minor :)

Thanks for following up about this so quickly, @ciampo! I see that https://github.com/WordPress/gutenberg/pull/47117 is merged, and I should be able to rebase and resolve conflicts within the next few days.

I will mark this as ready for review and ping you again soon!

**Comment by mike-day** [comment]
Regardless of a few failing mobile tests, I believe this PR is at a decent starting point if you want to take a look, @ciampo. Thanks!

File: packages/components/src/toolbar/stories/index.tsx
Comment (mirka) [inline packages/components/src/toolbar/stories/index.tsx]:
```suggestion
export const Default = Template.bind( {} );
```

Just a tip: These stories don't need type annotations as long as `Template` is typed 😎

File: packages/components/src/toolbar/stories/index.tsx
Comment (mirka) [inline packages/components/src/toolbar/stories/index.tsx]: Instead of `unknown`, should this also be a `@ts-expect-error` for when `ToolbarItem`/`DropdownMenu` is typed?

File: packages/components/src/toolbar/stories/index.tsx
Comment (mirka) [inline packages/components/src/toolbar/stories/index.tsx]: We generally prefer `@ts-expect-error` over a simple `@ts-ignore` so it's flagged when it's no longer an error. It helps prevent any outdated ignore statements from accumulating.

```suggestion
					// @ts-expect-error TODO: Remove when ToolbarGroup is typed
```

File: packages/components/src/toolbar/toolbar/types.ts
Comment (mirka) [inline packages/components/src/toolbar/toolbar/types.ts]: I don't quite think we should be adding `ReakitToolbarContainerProps` here, because a lot of them aren't intended for the consumer of the wp-components Toolbar to provide. They are intended for Reakit's `useToolbarState` hook to provide. There may be _some_ props that are legitimately useful, but there's also a benefit to controlling our own API surface in an intentional way so we don't lock ourselves into the Reakit implementation.

What do you think about keeping `ToolbarProps` really simple, like maybe `label` and `children` only, and just add in generic `div` props to be passed through? And we can always add more official props to the documentation when necessary.

```tsx
// toolbar/index.tsx
function UnforwardedToolbar(
	{
		className,
		label,
		...props
	}: WordPressComponentProps< ToolbarProps, 'div', false >, // 👈

// toolbar-container.tsx	
function UnforwardedToolbarContainer(
	{ label, ...props }: WordPressComponentProps< ToolbarProps, 'div', false >, // 👈
	ref: ForwardedRef< any >
```

**Review by mirka — COMMENTED** [review]
Thanks for contributing! 🙌 Good call to keep the scope of this PR to a manageable size.

I just had a few comments, but on the whole this looks clean and pretty close to merging!

File: packages/components/tsconfig.json
Comment (ciampo) [inline packages/components/tsconfig.json]: Nitty, but should we add the toolbar subfolders that are not typed yet here, instead of `@ts-ignore` at the top of the files?

Comment (mirka) [reply packages/components/tsconfig.json]: If they're not going to be typed in the next week or so, I think I prefer the `ts-nocheck`s so we can clear out this exclude list and ship the package types 🙈

Comment (ciampo) [reply packages/components/tsconfig.json]: Ah, I see! But wouldn't it be weird to ship partial types for `Toolbar`?

File: packages/components/src/toolbar/stories/index.tsx
Comment (mike-day) [reply packages/components/src/toolbar/stories/index.tsx]: Great catch! I just missed that one even though the typing for `Template` was right there.

File: packages/components/src/toolbar/stories/index.tsx
Comment (mike-day) [reply packages/components/src/toolbar/stories/index.tsx]: Got it, thanks for the tip! I will replace ignores with `@ts-expect-error` wherever possible.

File: packages/components/src/toolbar/toolbar/types.ts
Comment (mike-day) [reply packages/components/src/toolbar/toolbar/types.ts]: This is much cleaner, @mirka! Great suggestion, covered in [2cf63de](https://github.com/WordPress/gutenberg/pull/47087/commits/2cf63de78e044147e5fc5fb74842521aac224b19).

File: packages/components/tsconfig.json
Comment (mike-day) [reply packages/components/tsconfig.json]: Fwiw, I am starting work on `toolbar-button` today with a general plan to run through the remaining toolbar subfolders this month 💻

File: packages/components/src/toolbar/stories/index.tsx
Comment (mike-day) [reply packages/components/src/toolbar/stories/index.tsx]: Agreed! Covered in [dcc7a3a](https://github.com/WordPress/gutenberg/pull/47087/commits/dcc7a3ad9f7474b1efede6e69615ce68d7617e0a).

File: packages/components/CHANGELOG.md
Comment (mirka) [inline packages/components/CHANGELOG.md]: One last thing — could you just move this line up to the "Unreleased" section again?

**Review by mirka — APPROVED** [review]
Looks great, thank you! We can merge once the changelog is tweaked 🚀

File: packages/components/CHANGELOG.md
Comment (mike-day) [reply packages/components/CHANGELOG.md]: Thanks @mirka! This is done with [4b2dbd0](https://github.com/WordPress/gutenberg/pull/47087/commits/4b2dbd0381fbfc6985878e0f51bbf8d9de7a7787).

**Pull request merged**
